### PR TITLE
Implement OIDC implicit flow.

### PR DESCRIPTION
**Pull request by jimabramson**
ECOM-2440

depends on https://github.com/edx/django-oauth2-provider/pull/22 - will update setup/requirements when that has merged.

@clintonb @nasthagiri  please review
@justinabrahms FYI

### CODE REVIEW

File: oauth2_provider/tests/test_authorization.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        returned during the implicit flow.
+        """
+        self.assertTrue(location.startswith(self.auth_client.redirect_uri + '#'))
+        return {k: v[0] for k, v in parse_qs(urlparse(location).fragment).items()}
+
+    def _get_access_token(self, scope):
+        """
+        DRY helper.  Find the access token generated during our auth request.
+        """
+        return provider.oauth2.models.AccessToken.objects.get(
```
Comment (rlucioni) [inline oauth2_provider/tests/test_authorization.py]: I would import `AccessToken` from `provider.oauth2.models` above and change this to 
```python
return AccessToken.objects.get(...)
```

File: oauth2_provider/tests/test_authorization.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        self.assertEqual(302, response.status_code)
+        auth_payload = self._get_auth_payload(response['Location'])
+
+        # check state
+        self.assertEqual(auth_payload['state'], 'some_state')
+
+        # check scope
+        self.assertEqual(set(auth_payload['scope'].split()), set(scope.split()))
+
+        expected_token = self._get_access_token(scope)
```
Comment (rlucioni) [inline oauth2_provider/tests/test_authorization.py]: Should this only be attempted if `response_type == 'id_token token'`?

File: oauth2_provider/views.py [hunk 3]
```diff
@@ -11,7 +11,8 @@
 import provider.oauth2.forms
 import provider.scope
 from provider.oauth2.models import AccessToken
-from provider.oauth2.views import OAuthError, Capture, Redirect  # pylint: disable=unused-import
+from provider.oauth2.views import OAuthError, OAuth2AccessTokenMixin
+from provider.oauth2.views import Capture, Redirect  # pylint: disable=unused-import
```
Comment (rlucioni) [inline oauth2_provider/views.py]: Can you fold these imports together? They're all coming from `provider.oauth2.views`.

Comment (rlucioni) [inline oauth2_provider/views.py]: It'd be nice to clean up the other imports in here, too. Not necessary though.

File: oauth2_provider/constants.py [hunk 4]
```diff
@@ -41,6 +41,10 @@
 SCOPE_NAME_DICT = dict([(name, value) for (value, name) in SCOPES])
 SCOPE_VALUE_DICT = dict([(value, name) for (value, name) in SCOPES])
 
+# since this package explicitly supports the 'id_token' response type, override
+# the base package's default.
+DEFAULT_RESPONSE_TYPE_CHOICES = ('code', 'token', 'id_token')
+RESPONSE_TYPE_CHOICES = getattr(settings, 'OAUTH_RESPONSE_TYPE_CHOICES', DEFAULT_RESPONSE_TYPE_CHOICES)
```
Comment (nasthagiri) [inline oauth2_provider/constants.py]: Where is the base package's default being "overridden"?
Won't it not be updated if `'OAUTH_RESPONSE_TYPE_CHOICES'` is already set?

I would have expected this code to simply add `id_token` to the existing `provider.constants.RESPONSE_TYPE_CHOICES`.

File: oauth2_provider/views.py [hunk 5]
```diff
@@ -22,8 +23,103 @@
...
+        extra_data['id_token'] = self.encode_id_token(id_token)
+        scope = provider.scope.to_int(*id_token.scopes)
+
+        # Update the token scope, so it includes only authorized values.
+        access_token.scope = scope
+        access_token.save()
+
+        # Get the main fields for OAuth2 response.
+        response_data = super(OIDCTokenMixin, self).access_token_response_data(access_token, response_type)
+        if response_type and not response_type.endswith(' token'):
```
Comment (rlucioni) [inline oauth2_provider/views.py]: This seems brittle, but I may be missing something. If included, must `' token'` always be at the end of `response_type`? Why not check `'token' in response_type.split()`?

File: requirements.txt [hunk 6]
```diff
@@ -1,3 +1,5 @@
 django>=1.8.7
-edx-django-oauth2-provider>=0.3.0,<1.0.0
+# edx-django-oauth2-provider>=0.3.0,<1.0.0
+# TODO update with new django-oauth2-provider release
+-e git+https://github.com/edx/django-oauth2-provider.git@d6bcdd9de8b3e0620b1bd7273ff5a3fe17fa80bd#egg=django-oauth2-provider
```
Comment (rlucioni) [inline requirements.txt]: I assume this editable install is only here until edx/django-oauth2-provider#22 lands?

File: setup.py [hunk 7]
```diff
@@ -20,7 +20,8 @@
     ],
     packages=find_packages(exclude=['tests']),
     install_requires=[
-        'edx-django-oauth2-provider>=0.3.0,<1.0.0',
+        # TODO update with new django-oauth2-provider release
+        # 'edx-django-oauth2-provider>=0.3.0,<1.0.0',
```
Comment (rlucioni) [inline setup.py]: Likewise, I imagine this will be updated once edx/django-oauth2-provider#22 lands. I've commented there about updating the version number for that package so you can ensure it's installed here.

File: oauth2_provider/tests/test_authorization.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        ('openid profile', 'id_token'),
+        ('openid profile email', 'id_token'),
+    )
+    @ddt.unpack
+    def test_oidc_implicit_flow(self, scope, response_type):
+        """
+        Ensure that valid permutations of scope and response_type result in
+        correct OIDC behavior under the implicit flow.
+        """
+        response = self.login_and_authorize(scope=scope, response_type=response_type, trusted=True)
```
Comment (nasthagiri) [inline oauth2_provider/tests/test_authorization.py]: Shouldn't `trusted` be `False` for the implicit flow?

File: oauth2_provider/tests/test_authorization.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+            self.auth_client.client_secret,
+            audience=self.auth_client.client_id,
+        )
+
+    def test_oauth_implicit_flow(self):
+        """
+        Ensure that standard implicit flow takes effect if response type
+        'token' is used without the 'openid' scope.
+        """
+        response = self.login_and_authorize(scope='profile', response_type='token', trusted=True)
```
Comment (nasthagiri) [inline oauth2_provider/tests/test_authorization.py]: Same question about `trusted` here.

File: oauth2_provider/views.py (hunk 3 above)
Comment (jimabramson) [inline oauth2_provider/views.py]: the reason i separated the imports from `provider.oauth2.views` is because `Capture` and `Redirect` are imported from this module by (something...tests break without them) but not used locally.  So I only wanted the pylint directive to apply to those.

File: oauth2_provider/constants.py (hunk 4 above)
Comment (jimabramson) [inline oauth2_provider/constants.py]: on the one hand the base package's defaults are established [here](https://github.com/edx/django-oauth2-provider/blob/edx/provider/constants.py#L13) - since neither that package nor this one define their own runtime settings (instead it's done in the host django app, such as edx-platform) this seems like it should be effective.

on the other hand it doesn't work, at least in edx-platform.  it's still necessary to explicitly define the setting in platform, otherwise the base package's value defeats this one.  I think this is because the base package's constants end up being imported first.  and because of that, it almost seems not to matter what default is being set here.

i will fix this to append `id_token` as you'v'e suggested.

(btw is the correct term "overrided"?)

File: oauth2_provider/views.py (hunk 5 above)
Comment (jimabramson) [inline oauth2_provider/views.py]: hm...oidc docs imply 'token' will always come after 'id_token' but per rfc6749 the order should not matter, so i will fix this to be order-neutral.

File: requirements.txt (hunk 6 above)
Comment (jimabramson) [inline requirements.txt]: yes, holding off on the dependent PR (which has now merged, so will update)

File: oauth2_provider/tests/test_authorization.py (hunk 8 above)
Comment (jimabramson) [inline oauth2_provider/tests/test_authorization.py]: Our use case is single sign-on, where trusted is the right fit (we consider a logged-in edx user to implicitly authorize access to resources hosted on other federated domains).  We don't want an interstitial to be displayed when a new JWT is requested.

File: oauth2_provider/tests/test_authorization.py (hunk 2 above)
Comment (jimabramson) [inline oauth2_provider/tests/test_authorization.py]: in the test we need to fetch the underlying access token even if we don't expect it to be included in the response.   whether or not it is included in the response is still checked below (L61-66)

Comment (jimabramson) [inline oauth2_provider/tests/test_authorization.py]: wait - aha, that's not true anymore.  it used to be in a previous revision of this code.  my mistake - will update.

File: oauth2_provider/constants.py (hunk 4 above)
Comment (jimabramson) [inline oauth2_provider/constants.py]: well, after a second look i'm inclined to leave this as is.  appending 'id_token' will not have any different effect on the value in this module than explicitly defining the tuple value we want.  import issues in edx-platform can be fixed there.  again, whatever the host django app defines for `OAUTH_RESPONSE_TYPE_CHOICES` should take precedence anyway.

Comment (nasthagiri) [inline oauth2_provider/constants.py]: Ok.  I suppose we can revisit this if it becomes an issue.

> btw is the correct term "overrided"?
tomato-tomato

File: oauth2_provider/tests/test_authorization.py (hunk 8 above)
Comment (nasthagiri) [inline oauth2_provider/tests/test_authorization.py]: Ok. I now understand.  I confused this parameter with the standard [CLIENT_TYPES.PUBLIC constant](https://github.com/caffeinehit/django-oauth2-provider/blob/master/provider/constants.py#L7-L10).  I now see that this is our own custom [TrustedClient](https://github.com/edx/edx-oauth2-provider/blob/master/oauth2_provider/models.py#L15) model.

Comment (nasthagiri) [inline oauth2_provider/tests/test_authorization.py]: Note that the default `client_type` set in [ClientFactory](https://github.com/edx/edx-oauth2-provider/blob/master/oauth2_provider/tests/factories.py#L35) is CONFIDENTIAL.  However, for the implicit grant, we want to use and test with the PUBLIC `client_type`.

File: oauth2_provider/views.py (hunk 3 above)
Comment (rlucioni) [inline oauth2_provider/views.py]: Got it. It may be worth adding a comment about that inline.

File: tests/settings.py [hunk 10]
```diff
@@ -93,3 +93,5 @@
     'oauth2_provider.oidc.handlers.EmailHandler',
     'oauth2_provider.tests.handlers.TestHandler'
 )
+
+OAUTH_RESPONSE_TYPE_CHOICES = ('code', 'token', 'id_token')
```
Comment (nasthagiri) [inline tests/settings.py]: Is this change still needed, given https://github.com/edx/edx-oauth2-provider/pull/34/files#diff-1da467622f70e7cdcaee448b7cd6c981?

File: oauth2_provider/constants.py (hunk 4 above)
Comment (rlucioni) [inline oauth2_provider/constants.py]: The [past participle of "override" is "overridden."](https://www.google.com/search?q=overridden)

File: requirements.txt [hunk 11]
```diff
@@ -1,3 +1,2 @@
 django>=1.8.7
-edx-django-oauth2-provider>=0.3.0,<1.0.0
-PyJWT>=1.4.0,<2.0.0
+-e .
```
Comment (rlucioni) [inline requirements.txt]: Why does this package install itself as editable?

File: oauth2_provider/views.py [hunk 12]
```diff
@@ -22,8 +21,103 @@
...
+        # scopes, we cannot check if `openid` is the first scope to be
+        # requested, as required by OpenID Connect specification.
+
+        if not provider.scope.check(constants.OPEN_ID_SCOPE, access_token.scope):
+            if response_type and 'id_token' in response_type.split():
+                # http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest:
+                #   If the openid scope value is not present, the behavior is entirely unspecified.
+                #
+                # For our purposes, we'll treat this as a plain OAuth2 request and fall
+                # back to the default behavior, *unless* the 'id_token' scope is present,
```
Comment (nasthagiri) [inline oauth2_provider/views.py]: Should it be `'id_token' response_type` instead of `'id_token' scope`?

File: oauth2_provider/tests/test_authorization.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+            user=self.user, client=self.auth_client, scope=provider.scope.to_int(*scope.split())
+        )
+
+    @ddt.data(
+        ('openid', 'id_token token'),
+        ('openid', 'id_token'),
+        ('openid profile', 'id_token'),
+        ('openid profile email', 'id_token'),
+    )
+    @ddt.unpack
```
Comment (nasthagiri) [inline oauth2_provider/tests/test_authorization.py]: Do we support the other combinations that include `code`?  For example, the ones listed in:
http://openid.net/specs/oauth-v2-multiple-response-types-1_0.html#Combinations

If not, should we validate that they are not requested?

**Comment by nasthagiri** [comment]
@jimabramson Let me know if/when this PR is needed to be reviewed again.  Thanks.

**Comment by jimabramson** [comment]
@clintonb @nasthagiri I don't think we have any use case for implicit flow at this point - any reason for me not to trash this PR / branch ?

**Comment by clintonb** [comment]
Go for it. We can restore in the future, if necessary.
